gobblin-modules/google-ingestion/src/main/java/gobblin/ingestion/google/webmaster/GoogleWebmasterDataFetcherImpl.java

```diff
@@ -0,0 +1,366 @@
+                    _client.getPages(_siteProperty, startDate, endDate, country, GoogleWebmasterClient.API_ROW_LIMIT,
+                        requestedDimensions, apiDimensionFilters, start);
+                if (pages.size() < GoogleWebmasterClient.API_ROW_LIMIT) {
+                  //Figured out the size
+                  return pages.size() + start;
+                } else {
+                  //The size is not determined.
+                  return -1;
+                }
+              } catch (IOException e) {
+                LOG.info(String.format("Getting page size from %s failed. Retrying...", start));
+              }
+              try {
+                Thread.sleep(200);
+              } catch (InterruptedException e) {
+                LOG.error(e.getMessage());
+              }
+            }
+          }
+        });
+        results.add(submit);
+        try {
+          Thread.sleep(250);
+        } catch (InterruptedException e) {
+          LOG.error(e.getMessage());
+        }
+      }
+
+      for (Future<Integer> result : results) {
+        try {
+          Integer integer = result.get(2, TimeUnit.MINUTES);
+          if (integer > 0) {
+            es.shutdownNow();
+            return integer;
+          }
+        } catch (InterruptedException e) {
+          throw new RuntimeException(e);
+        } catch (ExecutionException e) {
+          throw new RuntimeException(e);
+        } catch (TimeoutException e) {
+          throw new RuntimeException(e);
+        }
+      }
+    }
+    throw new RuntimeException("Something seems wrong here. Having more than 5000*4*50 pages?");
+  }
+
+  /**
+   * Get all pages in an async mode.
+   */
+  private Collection<String> getPages(String startDate, String endDate, List<Dimension> dimensions,
+      ApiDimensionFilter countryFilter, Queue<Pair<String, FilterOperator>> toProcess) throws IOException {
+    String country = GoogleWebmasterFilter.countryFilterToString(countryFilter);
+
+    ConcurrentLinkedDeque<String> allPages = new ConcurrentLinkedDeque<>();
+    Random random = new Random();
+    //we need to retry many times because the path may be very long
+    final int retry = 120;
+    int r = 0;
+    while (r <= retry) {
+      ++r;
+      LOG.info(String.format("Get pages at round %d with size %d.", r, toProcess.size()));
+      ConcurrentLinkedDeque<Pair<String, FilterOperator>> nextRound = new ConcurrentLinkedDeque<>();
+      ExecutorService es = Executors.newFixedThreadPool(10, new ThreadFactory() {
+        @Override
+        public Thread newThread(Runnable r) {
+          Thread t = Executors.defaultThreadFactory().newThread(r);
+          t.setDaemon(true);
+          return t;
+        }
+      });
+
+      while (!toProcess.isEmpty()) {
+        submitJob(toProcess.poll(), countryFilter, startDate, endDate, dimensions, es, allPages, nextRound);
+        try {
+          Thread.sleep(275); //Submit roughly 4 jobs per second.
```

Comment:
This is not a google-mandated limit, but a good number that I realized during testing. It can suppress a lot of retries.